From d63803b51f21b63b43ed11644bd55a6a5489ee62 Mon Sep 17 00:00:00 2001
From: Valentina Fernandez <valentina.fernandezalanis@microchip.com>
Date: Mon, 28 Mar 2022 14:20:53 +0100
Subject: [PATCH] icicle: add support fot booting from QSPI flash memory

Updates to support booting from QSPI. Boot sequence is as follows:

1. Try to boot from an external QSPI flash memory device
2. Try to boot from SD/mmc
3. Try to boot from TFTP
---
 configs/microchip_mpfs_icicle_defconfig |  4 ++++
 include/configs/microchip_mpfs_icicle.h | 11 ++++++++++-
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/configs/microchip_mpfs_icicle_defconfig b/configs/microchip_mpfs_icicle_defconfig
index fd0c3dd47c..0bf890e981 100644
--- a/configs/microchip_mpfs_icicle_defconfig
+++ b/configs/microchip_mpfs_icicle_defconfig
@@ -15,3 +15,7 @@ CONFIG_SYS_PROMPT="RISC-V # "
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
 CONFIG_DM_MTD=y
+CONFIG_CMD_MTDPARTS=y
+CONFIG_MTDPARTS_DEFAULT="mtdparts=spi-nand0:8m(payload),128k(env),48m(fitimage)"
+CONFIG_USE_BOOTCOMMAND=y
+CONFIG_BOOTCOMMAND="setenv mmcbootpart 1; setenv scriptname boot.scr.uimg; run nandbootenv;run mmcbootenv; run dhcpbootenv"
diff --git a/include/configs/microchip_mpfs_icicle.h b/include/configs/microchip_mpfs_icicle.h
index 0b7e56af93..37c0864f92 100644
--- a/include/configs/microchip_mpfs_icicle.h
+++ b/include/configs/microchip_mpfs_icicle.h
@@ -31,6 +31,15 @@
 	"scriptaddr=0x1008100000\0" \
 	"pxefile_addr_r=0x1008200000\0" \
 	"ramdisk_addr_r=0x1008300000\0" \
-	BOOTENV
+	MPFS_BOOTENV
+
+#define MPFS_BOOTENV \
+	"mmcbootenv=echo Trying to boot from eMMC/SD...; if mmc rescan; then " \
+			"load mmc 0:${mmcbootpart} ${scriptaddr} ${scriptname}; source ${scriptaddr}; " \
+			"fi\0 " \
+	"nandbootenv=echo Trying to boot from QSPI...; if mtd list; then setenv mtd_present true ; mtd read env ${scriptaddr} 0; " \
+			"source ${scriptaddr}; setenv mtd_present; " \
+			"fi\0 " \
+	"dhcpbootenv=echo Trying to boot from DHCP/TFTP...; if dhcp ${scriptaddr} ${scriptname}; then source ${scriptaddr}; fi\0" \
 
 #endif /* __CONFIG_H */
-- 
2.25.1
